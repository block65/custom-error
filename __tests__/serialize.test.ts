import assert from 'node:assert';
import { test } from '@jest/globals';
import { CustomError, serializeError } from '../lib/index.js';

function throwUrlError() {
  return new URL('/', 'lol');
}

function throwAssertionError() {
  assert(false, 'asserting false');
}

[0, 'string', BigInt(0), false, Symbol('kaboom')].forEach((primitive) => {
  test('dont explode on primitives', () => {
    const result = serializeError(primitive);
    expect(result).toMatchSnapshot({
      stack: expect.any(String),
    });
  });
});

[null, undefined].forEach((primitive) => {
  test('dont explode on voids', () => {
    const result = serializeError(primitive);
    expect(result).toMatchSnapshot({
      stack: expect.any(String),
    });
  });
});

test('AssertionError', async () => {
  const serialized = await new Promise(throwAssertionError).catch(
    serializeError,
  );

  // WARN: this is serialized as a unrecognisable Error, due to Jest not
  // recognising the error as being instanceof Error
  // we cannot use isNativeError to check as it would no longer be isomorphic
  expect(serialized).toMatchSnapshot({
    stack: expect.any(String),
  });
});

test('URLError', async () => {
  const serialized = await new Promise(throwUrlError).catch(serializeError);

  // WARN: this is serialized as a unrecognisable Error, due to Jest not
  // recognising the error as being instanceof Error
  // we cannot use isNativeError to check as it would no longer be isomorphic
  expect(serialized).toMatchSnapshot({
    stack: expect.any(String),
  });
});

test('CustomError', () => {
  const previousErr = new CustomError('previous').debug({ woo1: 'woo1' });
  const current = new CustomError('current', previousErr).debug({
    woo2: 'woo2',
  });

  expect(current.cause).toBe(previousErr);

  const result = serializeError(current);

  expect(result).toMatchSnapshot({
    cause: expect.arrayContaining<Partial<SerializedError>>([
      expect.objectContaining<Partial<SerializedError>>({
        stack: expect.any(String) as string,
      }) as Partial<SerializedError>,
    ]) as Partial<SerializedError>[],
    stack: expect.any(String) as string,
  });
});
