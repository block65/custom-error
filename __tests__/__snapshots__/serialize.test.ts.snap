// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssertionError 1`] = `
Object {
  "debug": Object {
    "err": [AssertionError: asserting false],
    "typeofErr": "object",
  },
  "message": "AssertionError [ERR_ASSERTION]: asserting false",
  "name": "Error",
  "stack": Any<String>,
}
`;

exports[`CustomError 1`] = `
Object {
  "cause": ArrayContaining [
    ObjectContaining {
      "stack": Any<String>,
    },
  ],
  "code": 2,
  "debug": Object {
    "woo2": "woo2",
  },
  "message": "current",
  "name": "Error",
  "stack": Any<String>,
  "status": "UNKNOWN",
}
`;

exports[`URLError 1`] = `
Object {
  "debug": Object {
    "err": [TypeError: Invalid URL],
    "typeofErr": "object",
  },
  "message": "TypeError [ERR_INVALID_URL]: Invalid URL",
  "name": "Error",
  "stack": Any<String>,
}
`;

exports[`dont explode on primitives 1`] = `
Object {
  "debug": Object {
    "err": 0,
    "typeofErr": "number",
  },
  "message": "0",
  "name": "Error",
  "stack": Any<String>,
}
`;

exports[`dont explode on primitives 2`] = `
Object {
  "debug": Object {
    "err": "string",
    "typeofErr": "string",
  },
  "message": "string",
  "name": "Error",
  "stack": Any<String>,
}
`;

exports[`dont explode on primitives 3`] = `
Object {
  "debug": Object {
    "err": 0n,
    "typeofErr": "bigint",
  },
  "message": "0",
  "name": "Error",
  "stack": Any<String>,
}
`;

exports[`dont explode on primitives 4`] = `
Object {
  "debug": Object {
    "err": false,
    "typeofErr": "boolean",
  },
  "message": "false",
  "name": "Error",
  "stack": Any<String>,
}
`;

exports[`dont explode on primitives 5`] = `
Object {
  "debug": Object {
    "err": Symbol(kaboom),
    "typeofErr": "symbol",
  },
  "message": "Symbol(kaboom)",
  "name": "Error",
  "stack": Any<String>,
}
`;

exports[`dont explode on voids 1`] = `
Object {
  "debug": Object {
    "err": null,
    "typeofErr": "object",
  },
  "message": "null",
  "name": "Error",
  "stack": Any<String>,
}
`;

exports[`dont explode on voids 2`] = `
Object {
  "debug": Object {
    "err": undefined,
    "typeofErr": "undefined",
  },
  "message": "undefined",
  "name": "Error",
  "stack": Any<String>,
}
`;
